Crypto refactor

johningve opened the pull request:
This is a complete rewrite of the crypto implementations based on discussion in #34.

# Important Changes:

* The `CryptoImpl` interface is replaced with the `CryptoBase` interface which is much simpler:
```go
// CryptoBase provides the basic cryptographic methods needed to create, verify, and combine signatures.
type CryptoBase interface {
	// Sign creates a cryptographic signature of the given message.
	Sign(message []byte) (signature QuorumSignature, err error)
	// Combine combines multiple signatures into a single signature.
	Combine(signatures ...QuorumSignature) (signature QuorumSignature, err error)
	// Verify verifies the given quorum signature against the message.
	Verify(signature QuorumSignature, message []byte) bool
	// BatchVerify verifies the given quorum signature against the batch of messages.
	BatchVerify(signature QuorumSignature, batch map[hotstuff.ID][]byte) bool
}
```
* The `ThresholdSignature` interface is renamed to `QuorumSignature`, and the `Signature` interface is removed.
* `ConnectionMetadata` allows for metadata to be sent to other replicas.
* This metadata is used by the `bls12` package to send a proof-of-possession to all replicas.

Raytar commented:
@meling There are some changes to various tests because I had to change the `testutil.TestModules` function and make the `testutil.CreateBuilders` use the twins network to get the new metadata working. It's kinda ugly now, but I'll probably revisit it at some point and try to clean it up.

meling commented on backend/config.go:
```diff
@@ -110,6 +116,50 @@ func NewConfig(creds credentials.TransportCredentials, opts ...gorums.ManagerOpt
+func readMetadata(md metadata.MD) map[string]string {
+	m := make(map[string]string)
+	for k, values := range md {
+		if strings.HasPrefix(k, keyPrefix) && len(values) > 0 {
```
Go 1.18 has a new `strings.Cut` function that could replace both `HasPrefix` and `TrimPrefix`. Something like this, I think:
```go
if _, key, ok := strings.Cut(k, keyPrefix); ok && len(values) > 0 {
  m[key] = values[0]
```
Maybe the most important place to do this, but if done elsewhere as well, maybe `Cut` can be more efficient.

meling replied:
Suggest to use `Cut`, since we probably want to switch go Go 1.18 soon anyway.

meling commented on backend/config.go:
```diff
@@ -75,17 +84,14 @@ type Config struct {
+			return
+		}
+		c := event.(replicaConnected)
+		cfg.replicaConnected(c)
```
```go
cfg.replicaConnected(event.(replicaConnected))
```

meling commented on backend/server.go:
```diff
@@ -35,9 +34,11 @@ func (srv *Server) InitConsensusModule(mods *consensus.Modules, _ *consensus.Opt
-
-	opts = append(opts, gorums.WithGRPCServerOptions(grpcServerOpts...))
+	// grpcServerOpts := []grpc.ServerOption{}
+	// opts = append(opts, gorums.WithGRPCServerOptions(grpcServerOpts...))
```
Are these two lines no longer necessary?

meling commented on consensus/cryptoiface.go:
```diff
@@ -0,0 +1,82 @@
+
+// VerifyOptions specify options for the Verify function in the CryptoBase interface.
+type VerifyOptions struct {
+	// using a pointer for the Message such that we can tell if it is unspecified
```
I think `Message` does not need to be a pointer. You can still distinguish these two cases, if needed:
```go
b = []byte{} // non-nil, but len(b) == 0
b = nil // nil and len(b) == 0
```
I tried to remove the pointer and update the relevant code and the tests in the `crypto` package still pass.

meling commented on consensus/cryptoiface.go:
```diff
@@ -0,0 +1,82 @@
+
+// Crypto implements the methods required to create and verify signatures and certificates.
+// This is a higher level interface that is implemented by the crypto package itself.
+type Crypto interface {
```
I'm not a fan of big interfaces like this. Any reason to have all these methods in the same interface?

If you make separate interfaces for each method or pairs of methods, it would be easier to implement alternatives. Although, I'm not sure we will have multiple implementations of these methods.

Raytar replied:
The reason we have this interface is to avoid an import cycle. The only implementor is the `crypto` struct in `crypto/crypto.go`, and I don't see any reason for alternative implementations. But the methods of the `crypto` struct are needed by the other modules. So we can have this interface, or we would have to move the whole `crypto` struct into the `consensus` package as well. Maybe there is a better way, but I haven't been able to think of one yet.

meling commented on consensus/types.go:
```diff
@@ -110,33 +103,43 @@ type PrivateKey interface {
 // PartialCert is a signed block hash.
 type PartialCert struct {
-	signature Signature
+	// shortcut to the signer of the signature
```
In general, I'm not sure the `consensus` package is the right place for these certificate types. I wonder if it wouldn't be better to place them in `crypto` or `crypto/certificates` or something along these lines.

meling replied:
Marking this as resolved since I created an issue for the possible refactor of the module system.

meling commented on consensus/cryptoiface.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+// CryptoBase provides the basic cryptographic methods needed to create, verify, and combine signatures.
+type CryptoBase interface {
```
See the naming/composition here: [`Signer`](https://pkg.go.dev/crypto#Signer). There is no interface that includes a verify function. On the other hand, it seems the std lib doesn't define an interface that's used across signature types. Instead most signing/verify algorithms just use functions specific for each signature type.

meling replied:
I guess my point was that I don't think `CryptoBase` is the right name for this interface, since the methods are related to digital signatures in some form. The other question is whether or not we should use an interface.

meling replied:
Similar to other question: why is this in `consensus`?

Raytar replied:
I haven't yet found a better place to put these types/interfaces without import cycles. It would be best to keep them in the crypto package, but I also need the interfaces to be available in the module system. And the crypto package needs the module system...

meling replied:
We should ideally refactor things to avoid such strong coupling, but that should be done as a separate PR.

meling replied:
Marking this as resolved since I created an issue for the possible refactor of the module system.

meling commented on backend/config.go:
```diff
@@ -110,6 +115,50 @@ func NewConfig(creds credentials.TransportCredentials, opts ...gorums.ManagerOpt
+
+const keyPrefix = "hotstuff-"
+
+func convertMetadata(m map[string]string) metadata.MD {
```
Maybe rename this to `mapToMetadata`

meling commented on backend/config.go:
```diff
@@ -119,6 +168,18 @@ type ReplicaInfo struct {
 // Connect opens connections to the replicas in the configuration.
 func (cfg *Config) Connect(replicas []ReplicaInfo) (err error) {
+	opts := *cfg.optsPtr
+	cfg.optsPtr = nil // we don't need to keep the options around beyond this point, so we'll allow them to be GCed.
```
Is this an important point? I mean is this causing memory leaks or something? Maybe it would be better to make a copy of options and let that be GCed when exiting the method. I mean `Connect` should be called rarely, right?

meling commented on backend/config.go:
```diff
@@ -150,6 +212,9 @@ func (cfg *Config) Connect(replicas []ReplicaInfo) (err error) {
 	}
 
+	cfg.connected = true
+	cfg.mods.EventLoop().AddEvent(connected{})
```
Would be nice with a comment referring back to the `connected{}` event handler, and what it will do on this event.

meling commented on consensus/cryptoiface.go:
```diff
@@ -0,0 +1,82 @@
+// VerifySingle verifies the signature against the given message.
+//
+// NOTE: VerifySingle and VerifyMulti cannot be used together,
+// and whichever option is specified last will be the effective one.
```
If they cannot be used together, would it not be better to panic, since it would be a programmer mistake to combine them??

meling commented on consensus/cryptoiface.go:
```diff
@@ -0,0 +1,82 @@
+
+	Message          *[]byte
+	MessageMap       map[hotstuff.ID][]byte
+	MultipleMessages bool // true if `MessageMap` should be used instead of `Message`
```
Couldn't we encode the single message variant in the `MessageMap` with `hotstuff.ID = 0`?? Then we could avoid the `Message` field and probably the `MultipleMessages` field as well.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
-	bc := &bls12Crypto{}
-	return bc
+	mut      sync.RWMutex
+	popCache map[string]bool
```
What is `popCache` storing? Add comment?

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
-	pk := bc.getPrivateKey()
-	bls12.NewG2().MulScalarBig(p, p, pk.p)
-	return &Signature{signer: bc.mods.ID(), s: p}, nil
+func (bls *bls12Base) getPrivateKey() *PrivateKey {
```
`s/getPrivateKey/privateKey/g`??

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+		}
+	}()
+
+	b, ok := replica.Metadata()[popMetadataKey]
```
What is `b`? A more descriptive variable name would help.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+		return valid
+	}
+
+	p, err := bls12.NewG2().FromCompressed([]byte(b))
```
What is `p`? More descriptive variable name.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -86,35 +91,6 @@ func (priv *PrivateKey) Public() consensus.PublicKey {
-	return s.signer
-}
-
 // AggregateSignature is a bls12-381 aggregate signature. The participants map contains the IDs of the replicas who
```
... The participants bitmap contains... the replicas that participated in signature creation.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+	pubKey := bls.getPrivateKey().Public().(*PublicKey)
+	proof, err := bls.coreSign(pubKey.ToBytes(), domainPOP)
+	if err != nil {
+		bls.mods.Logger().Panicf("failed to generate proof of possession: %v", err)
```
Should logged messages be capitalized? `"Failed to ..."`

meling replied:
I assume this will panic. What's the policy for logging a panic vs emitting it to stdout; I assume this is the difference... There are other places where regular panic is used.

Raytar replied:
Actually, I don't have a policy for that.

Here are some more differences when compared with `panic`:

`logging.Panic` and `logging.Panicf`:
- can only be used where the logger module is accessible
- the message will be logged even if the panic is recovered
- `Panicf` supports string formatting

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+
+func (bls *bls12Base) coreAggregateVerify(publicKeys []*PublicKey, messages [][]byte, signature *bls12.PointG2) bool {
+	n := len(publicKeys)
+	if n != len(messages) && n <= 1 {
```
Do you need `&& n <= 1`? Does this method not work for `n=1` and `len(publicKeys) == len(messages)`? Maybe it would help to add a comment to explain the limitations of this method.

Raytar replied:
Oops, this is actually wrong. n >= 1 is a precondition of the function, and the length of the public keys and messages slices must be equal

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+
+// Sign creates a cryptographic signature of the given messsage.
+func (bls *bls12Base) Sign(message []byte) (signature consensus.QuorumSignature, err error) {
+	// TODO: consider passing message to this function instead of hash
```
Still valid TODO? The input `message` seems to indicate it is already a message.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+	)
+
+	// get all public keys and messages to verify
+	sig.participants.RangeWhile(func(i hotstuff.ID) bool {
```
`s/i/id/g`

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
-		pk, ok := replica.PublicKey().(*PublicKey)
-		if !ok {
+
+		if replica.ID() != bls.mods.ID() && !bls.checkPop(replica) {
```
A comment on this line would help.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
-			return false
+		publicKeys = append(publicKeys, replica.PublicKey().(*PublicKey))
+
+		if opts.MultipleMessages {
```
If `MultipleMessages` is false, then `messages` above is not used.

Overall, it seems like the code could be simplified by avoiding the separation between single message verify functions and multi-message verify functions.

Raytar replied:
The problem with removing this separation, i.e. using the map for both `VerifySingle` and `VerifyMulti` is that it becomes difficult to check which option was used. For example, if the length is 1 and there is only one participant in the signature, then you still don't know if `VerifySingle` or `VerifyMulti` was used.

meling replied:
Okay.

meling commented on crypto/cache.go:
```diff
@@ -3,36 +3,27 @@ package crypto
 	accessOrder list.List
 }
 
 // NewCache returns a new Crypto implementation that caches the results of the operations of the given CryptoImpl
```
`s/CryptoImpl/CryptoBase/`

meling commented on crypto/cache.go:
```diff
@@ -193,7 +130,8 @@ func (cache *cache) VerifyThresholdSignatureForMessageSet(signature consensus.Th
+func (cache *cache) Combine(signatures ...consensus.QuorumSignature) (consensus.QuorumSignature, error) {
 	// we don't cache the result of this operation, because it is not guaranteed to be valid.
-	return cache.impl.Combine(signatures...)
+	sig, err := cache.impl.Combine(signatures...)
```
Since `err` isn't checked, you can simply return directly.

meling commented on crypto/cache.go:
```diff
@@ -73,117 +63,64 @@ func (cache *cache) evict() {
-// Sign signs a hash.
-func (cache *cache) Sign(hash consensus.Hash) (sig consensus.Signature, err error) {
-	sig, err = cache.impl.Sign(hash)
+// Sign signs a message.
```
... and adds it to the cache.

meling commented on crypto/cache.go:
```diff
@@ -73,117 +63,64 @@ func (cache *cache) evict() {
-	cache.insert(key, false)
+	var key strings.Builder
+	hash := sha256.Sum256(message)
+	_, _ = key.Write(hash[:])
```
Suggest to add a comment to explain why we ignore checking these errors...

Raytar replied:
It is because writing to a string builder never returns an error. It is just immplementing the io.Writer interface.

meling replied:
Reasonable. Probably fine to not add a comment then.

meling commented on crypto/cache.go:
```diff
@@ -73,117 +63,64 @@ func (cache *cache) evict() {
+	// if we have multiple messages, we will hash them all together.
+	if opts.MultipleMessages {
+		hasher := sha256.New()
+		for _, m := range opts.MessageMap {
```
Won't this result in non-deterministic iteration ordering?

meling commented on crypto/cache.go:
```diff
@@ -193,7 +130,8 @@ func (cache *cache) VerifyThresholdSignatureForMessageSet(signature consensus.Th
 // As opposed to the CreateThresholdSignature methods,
```
Check this documentation.. including ^^ Should it be `single threshold`?

meling commented on crypto/crypto.go:
```diff
@@ -6,112 +6,120 @@ import (
 }
 
-func (base base) CreateAggregateQC(view consensus.View, timeouts []consensus.TimeoutMsg) (aggQC consensus.AggregateQC, err error) {
+func (c crypto) CreateAggregateQC(view consensus.View, timeouts []consensus.TimeoutMsg) (aggQC consensus.AggregateQC, err error) {
```
Missing doc

meling replied:
Is there a linter that could check this. Probably `golangci-lint` has one, but didn't check.

Raytar replied:
We have a linter for that, but it doesn't care about this method, because it is defined on an unexported struct.

meling commented on crypto/crypto.go:
```diff
@@ -6,112 +6,120 @@ import (
-	sig, err := base.CreateThresholdSignature(sigs, view.ToHash())
+	sig, err := c.Combine(sigs...)
 	if err != nil {
 		return consensus.TimeoutCert{}, err
```
We should avoid returning empty structs when we have error; it causes unnecessary allocation.

meling commented on crypto/crypto.go:
```diff
@@ -6,112 +6,120 @@ import (
-	sig, err := base.CreateThresholdSignature(sigs, block.Hash())
+	sig, err := c.Combine(sigs...)
 	if err != nil {
 		return consensus.QuorumCert{}, err
```
We should avoid returning empty structs when we have error; it causes unnecessary allocation.

Raytar replied:
I thought that the memory would be allocated on the stack since it is returned by value?

meling replied:
I have gotten into the habit of always returning a pointer type in such scenarios. I'm not sure if it always doesn't escape to the heap, but it probably doesn't matter much in terms of performance. I had to look it up, and it seems both are okay, but it seems the [recommended pattern](https://morioh.com/p/385f6171e967) is to use the `nil` variant. This seems to be based on the advice on [Declaring Empty Slices](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices), which also references a talk about nil that I've watched before but didn't rewatch.

I guess one argument for using a `nil` pointer is that it becomes easier to identify/see that it is in fact returning a `nil` as opposed to examining the code and seeing `consensus.QuorumCert{}` you sort of have to think what it means that this is non-nil.

Raytar replied:
I just don't think that it is a good idea for constructor functions to return a pointer unless it is needed for some specific purpose. What good could a pointer do in this case?

meling commented on crypto/crypto.go:
```diff
@@ -6,112 +6,120 @@ import (
+func (c crypto) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
+	sig, err := c.Sign(block.ToBytes())
 	if err != nil {
 		return consensus.PartialCert{}, err
```
We should avoid returning empty structs when we have error; it causes unnecessary allocation.

meling commented on crypto/crypto.go:
```diff
@@ -6,112 +6,120 @@ import (
-	sig, err := base.CreateThresholdSignatureForMessageSet(sigs, hashes)
+	sig, err := c.Combine(sigs...)
 	if err != nil {
 		return aggQC, err
```
We should avoid returning non-nil structs when we have error...

meling commented on crypto/crypto.go:
```diff
@@ -6,112 +6,120 @@ import (
 // VerifyQuorumCert verifies a quorum certificate.
-func (base base) VerifyQuorumCert(qc consensus.QuorumCert) bool {
+func (c crypto) VerifyQuorumCert(qc consensus.QuorumCert) bool {
 	if qc.BlockHash() == consensus.GetGenesis().Hash() {
```
Perhaps add comment here:
```go
// genesis QC is always valid.
```

meling commented on crypto/crypto.go:
```diff
@@ -6,112 +6,120 @@ import (
 // VerifyTimeoutCert verifies a timeout certificate.
-func (base base) VerifyTimeoutCert(tc consensus.TimeoutCert) bool {
+func (c crypto) VerifyTimeoutCert(tc consensus.TimeoutCert) bool {
 	if tc.View() == 0 {
```
Perhaps add comment here too.

meling commented on crypto/crypto.go:
```diff
@@ -121,17 +129,17 @@ func (base base) VerifyAggregateQC(aggQC consensus.AggregateQC) (bool, consensus
-	ok := base.VerifyThresholdSignatureForMessageSet(aggQC.Sig(), hashes)
+	ok := c.Verify(aggQC.Sig(), consensus.VerifyMulti(messages))
 	if !ok {
 		return false, consensus.QuorumCert{}
```
Should probably return `nil, false`...

meling reviewed (changes requested):
Looks nice! Mainly, I would prefer to have the verification functions work on a slice or map instead of having to decide via a boolean which path to take; this would also avoid two fields when you only need one of them... if necessary, the bool could be replaced with a `len(x) == 1` check.

meling commented on crypto/bitfield.go:
```diff
@@ -9,18 +9,21 @@ type Bitfield struct {
-func (bm Bitfield) isSet(byteIdx, bitIdx int) bool {
-	return bm.data[byteIdx]&(1<<bitIdx) != 0
+func (bf Bitfield) isSet(byteIdx, bitIdx int) bool {
+	return bf.data[byteIdx]&(1<<bitIdx) != 0
```
The id and index methods below should use shift `>> 3` and `<< 3`. Although not sure if the compiler replaces `*8` and `/8` accordingly...

Raytar replied:
Why?

meling replied:
Well, I guess it is only necessary/useful if the compiler doesn't already replace Divide and Multiply operations with shift operations, which are substantially faster. Doing it ourselves makes it fast for sure, independent of compiler optimizations, but I guess it makes the code slightly less readable. (I guess a comment can alleviate the latter.)

meling replied:
On second thought, maybe it is better to leave it as is for clarity... For fun, I studied the assembly code generated by these operations. It is actually shorter code with the `/` and `*` operations, but only because of the `%` operation. That is, the simple `* 8` does become a shift `<< 3`. But the `/ 8` doesn't become only a shift operation, when compiled standalone. However, since the function in question has both `/ 8` and `% 8`, it seems to be able to reuse the parts that aren't shift... (but the whole thing is non-trivial to really understand...) The main takeaway is that compilers are smart and we probably should do less manual optimization unless it is demonstrably faster in benchmarks.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -318,12 +318,16 @@ func (bls *bls12Base) Verify(signature consensus.QuorumSignature, options ...con
-	if !opts.MultipleMessages && opts.Message == nil {
-		panic("no hash(es) to verify the signature against: you must specify one of the VerifyHash or VerifyHashes options")
+	if len(opts.Messages) == 0 {
+		panic("no message(s) to verify the signature against: you must specify one of the VerifyHash or VerifyHashes options")
```
`s/VerifyHash/VerifySingle`
`s/VerifyHashes/VerifyMulti`

meling commented:
We should make some progress towards getting this merged. What remains to be discussed and what remains to be implemented?

johningve commented:
@meling I split the Verify function into two `Verify` and `BatchVerify`. I think this is more sensible than the approach of using the `VerifyOption` stuff. These functions are no longer responsible for checking whether the number of signatures constitutes a quorum, I moved that to the higher-level `Crypto` module's functions.

meling commented on backend/config.go:
```diff
@@ -52,16 +55,22 @@ func (r *Replica) NewView(msg consensus.SyncInfo) {
-	mods    *consensus.Modules
-	optsPtr *[]gorums.ManagerOption // using a pointer so that options can be GCed after initialization
+	mods      *consensus.Modules
+	opts      []gorums.ManagerOption // using a pointer so that options can be GCed after initialization
```
This no longer uses a pointer to the slice; maybe the comment should be removed or rephrased?

meling commented on backend/config.go:
```diff
@@ -75,17 +84,16 @@ type Config struct {
-	md := metadata.New(map[string]string{
-		"id": fmt.Sprintf("%d", cfg.mods.ID()),
+	// This receives `replicaConnected` events from the server.
+	// We can only process these events after the configuration has been connected,
```
The comment was a bit longwinded. Suggest rephrasing, e.g.,:
```
We delay processing `replicaConnected` events until after the configuration's `connected` event has occurred.
```

meling commented on backend/config.go:
```diff
@@ -119,15 +171,28 @@ type ReplicaInfo struct {
 // Connect opens connections to the replicas in the configuration.
 func (cfg *Config) Connect(replicas []ReplicaInfo) (err error) {
+	opts := cfg.opts
+	cfg.opts = nil // we don't need to keep the options around beyond this point, so we'll allow them to be GCed.
```
```go
// options are not needed beyond this point, so we allow them to be GCed.
```

meling commented on crypto/bls12/bls12.go:
```diff
@@ -86,37 +91,8 @@ func (priv *PrivateKey) Public() consensus.PublicKey {
-
-// AggregateSignature is a bls12-381 aggregate signature. The participants map contains the IDs of the replicas who
-// participated in the creation of the signature. This allows us to build an aggregated public key to verify the
+// AggregateSignature is a bls12-381 aggregate signature. The participants bitmap contains the IDs of the replicas that
```
Just to clarify, do you mean `The participant's bitmap` or the plural form??

johningve replied:
I mean the `participants` field in the struct.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
-	bc := &bls12Crypto{}
-	return bc
+	mut sync.RWMutex
+	// popCache caches the results of popVerify for each public key.
```
Should mention here and probably other places that `pop` = `proof of possession`

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+	pubKey := bls.privateKey().Public().(*PublicKey)
+	proof, err := bls.coreSign(pubKey.ToBytes(), domainPOP)
+	if err != nil {
+		bls.mods.Logger().Panicf("Failed to generate proof of possession: %v", err)
```
Should be `proof-of-possession` or Lines 237 and 243 should be changed.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+func (bls *bls12Base) checkPop(replica consensus.Replica) (valid bool) {
+	defer func() {
+		if !valid {
+			bls.mods.Logger().Debugf("Invalid proof-of-possession for replica %d", replica.ID())
```
is this debug or warn?

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+			}
+			g2.Add(&agg, &agg, &sig2.sig)
+		} else {
+			bls.mods.Logger().Panicf("cannot combine signature of incompatible type %T (expected %T)", sig1, sig2)
```
`Cannot combine incompatible signature type ...` (same applies to Line 364.)

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+		bls.mods.Logger().Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
+	}
+
+	l := s.Participants().Len()
```
Previously `n` was used for a similar length. Suggest using `n` instead, because `l` looks almost like `1`.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+	// else:
+	pks := make([]*PublicKey, 0, l)
+	s.Participants().ForEach(func(id hotstuff.ID) {
+		pk, ok := bls.publicKey(id)
```
Can be moved into the `if` since `pk` isn't used later in the block.

johningve replied:
`ok` is checked by the if statement though.

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
-	if len(sigs) < bc.mods.Configuration().QuorumSize() {
-		return nil, multierr.Combine(crypto.ErrNotAQuorum, err)
+
+	// else:
```
Add more context; assume you referring to the `if l == 1`, meaning that `l > 1`.

johningve replied:
Yeah, will do. Here I would prefer to have an explicit else clause, because I think it makes the code clearer, but it goes against Go's style (and linters will complain).

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+func (bls *bls12Base) BatchVerify(signature consensus.QuorumSignature, batch map[hotstuff.ID][]byte) bool {
+	s, ok := signature.(*AggregateSignature)
+	if !ok {
+		bls.mods.Logger().Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
```
`Cannot verify incompatible signature type`

meling commented on crypto/bls12/bls12.go:
```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+		if ok {
+			pks = append(pks, pk)
+		} else {
+			bls.mods.Logger().Warnf("Missing public key for ID %d", id)
```
Does it make sense to continue the loop in this case? Ref. to a similar construction on Line 413.

johningve replied:
No it does not. Will change.

meling commented on crypto/cache.go:
```diff
@@ -73,127 +65,134 @@ func (cache *cache) evict() {
-// this method does not check whether the resulting
-// signature meets the quorum size.
-func (cache *cache) Combine(signatures ...interface{}) consensus.ThresholdSignature {
+// Verify verifies a signature given a hash.
```
Do we still need this commented code?

meling commented on crypto/cache.go:
```diff
@@ -3,36 +3,29 @@ package crypto
 }
 
-// NewCache returns a new Crypto implementation that caches the results of the operations of the given CryptoImpl
+// NewCache returns a new Crypto implementation that caches the results of the operations of the given CryptoBase
```
NewCache returns a new Crypto instance that caches the results of the operations of the provided CryptoBase.

meling commented on crypto/cache.go:
```diff
@@ -73,127 +65,134 @@ func (cache *cache) evict() {
-	if cache.check(key, true) {
+// BatchVerify verifies the given quorum signature against the batch of messages.
+func (cache *cache) BatchVerify(signature consensus.QuorumSignature, batch map[hotstuff.ID][]byte) bool {
+	// get a sorted list of the ids in the Messages map
```
```
// sort the list of ids from the batch map
```

meling commented on crypto/cache.go:
```diff
@@ -73,127 +65,134 @@ func (cache *cache) evict() {
+func (cache *cache) BatchVerify(signature consensus.QuorumSignature, batch map[hotstuff.ID][]byte) bool {
+	// get a sorted list of the ids in the Messages map
+	ids := make([]hotstuff.ID, 0, len(batch))
+	for id := range batch {
```
Can't recall having seen this construction before to sort the map keys, but it looks more efficient than to first loop over the map to create a slice before sorting. That said, I wonder if it would be better to make this a function called `SortedIDs` that takes a map with `hotstuff.ID` and possibly any type (when we use generics -- assuming we need to sort IDs from maps with other types.) 

For the record, there isn't coming a `SortedKeys` in the stdlib yet, but maybe in [the future.](https://github.com/golang/go/issues/53196)

johningve replied:
I played around with some different ways to make this generic. However, I think that that using the generic `maps.Keys` function from the experimental `maps` package to extract the keys to a slice, and then the generic `slices.Sort` function from the experimental `slices` package to sort the keys is easier and likely faster.

johningve replied:
Here are some benchmark results:

```
cpu: AMD Ryzen 9 5900HS with Radeon Graphics        
BenchmarkKeysAsc-16      	       4	 282953742 ns/op	 2238214 B/op	     979 allocs/op
BenchmarkSliceSort-16    	     104	  11535986 ns/op	  858424 B/op	      42 allocs/op
```

The first one, `KeysAsc`, uses the sort.Search method in the same way as the code above.
The second one, `SliceSort`, first uses `maps.Keys` and then sorts them with `slices.Sort`.

Both are sorting the keys of a map with 100 000 entries

johningve replied:
When sorting maps with only 10 entries, the `slices.Sort` version still wins:

```
cpu: AMD Ryzen 9 5900HS with Radeon Graphics        
BenchmarkKeysAsc-16      	 5160612	       230.5 ns/op	      80 B/op	       1 allocs/op
BenchmarkSliceSort-16    	 7871254	       152.8 ns/op	      80 B/op	       1 allocs/op
```

meling commented on crypto/cache.go:
```diff
@@ -73,127 +65,134 @@ func (cache *cache) evict() {
+	hasher := sha256.New()
+	// then hash the messages in sorted order
+	for _, id := range ids {
+		m := batch[id]
```
Can replace `m` with `batch[id]` directly in the `Write` call.

meling commented on crypto/crypto.go:
```diff
@@ -0,0 +1,153 @@
+
+// VerifyPartialCert verifies a single partial certificate.
+func (c crypto) VerifyPartialCert(cert consensus.PartialCert) bool {
+	block, ok := c.mods.BlockChain().Get(cert.BlockHash())
```
Wouldn't it be better to pass in `block` to the function, instead of accessing the blockchain in the crypto package, causing an implicit dependency?  Same applies to Line 109. Maybe this is fixed in #61?

johningve replied:
I guess it is a responsibility problem. Should the caller of `VerifyPartialCert` be responsible for finding the block, or should the crypto implementation itself do it? I prefer it the way it is because it makes the crypto interface easier to use.

meling commented on crypto/crypto.go:
```diff
@@ -0,0 +1,153 @@
+	if qc.BlockHash() == consensus.GetGenesis().Hash() {
+		return true
+	}
+	if qc.Signature().Participants().Len() < c.mods.Configuration().QuorumSize() {
```
Similar to my [comment above](https://github.com/relab/hotstuff/pull/36/files#r888374558), the quorum size or Configuration could be passed in, which would decouple the crypto package from the module system.

At least my initial reaction was that this may cause problems due to tight coupling.

johningve replied:
I think that passing in the quorum size or configuration complicates the interface. I don't think the Crypto module's dependency on the BlockChain module is a problem.

meling commented on crypto/crypto.go:
```diff
@@ -0,0 +1,153 @@
+	if aggQC.Sig().Participants().Len() < c.mods.Configuration().QuorumSize() {
+		return consensus.QuorumCert{}, false
+	}
+	ok = c.BatchVerify(aggQC.Sig(), messages)
```
Move into if?

meling replied:
Actually, this code (Lines 145-152) could be written as:
```go
if c.BatchVerify(aggQC.Sig(), messages) && c.VerifyQuorumCert(highQC) {
	return highQC, true
}
return consensus.QuorumCert{}, false
```
Perhaps adding a comment above the if to say that both must verify...

meling commented on crypto/ecdsa/ecdsa.go:
```diff
@@ -127,225 +123,146 @@ func (sig ThresholdSignature) RangeWhile(f func(hotstuff.ID) bool) {
-		if !ok {
-			continue
-		}
+	l := signature.Participants().Len()
```
`s/l/n/g`

meling replied:
The above comment meant: substitute `l` with `n`.

johningve replied:
Indeed, I am familiar with sed syntax ;)

meling commented on crypto/ecdsa/ecdsa.go:
```diff
@@ -127,225 +123,146 @@ func (sig ThresholdSignature) RangeWhile(f func(hotstuff.ID) bool) {
-	ec.mods.Logger().Debug(hashes)
-	sig, ok := signature.(ThresholdSignature)
-	if !ok {
+	l := signature.Participants().Len()
```
`s/l/n/g`

meling commented on crypto/ecdsa/ecdsa.go:
```diff
@@ -127,225 +123,146 @@ func (sig ThresholdSignature) RangeWhile(f func(hotstuff.ID) bool) {
 		if !ok {
 			return false
 		}
+		set[string(message)] = struct{}{}
```
This might be costly for long messages. Perhaps it would be better to stringify the `hash` and use that as the map key?? It would be interesting to benchmark this. @hanish520

johningve replied:
Actually, we can just use the hash itself. No need to stringify

meling reviewed (changes requested):
Looks great! There are a few minor things, and a few things that need some discussion. I don't want to hold it back longer, so the things to be discussed can also be postponed for another PR, if we decide they would help decoupling.

I noticed tests are failing due to golangci-lint, as you already noticed in #58; probably replacing [this line](https://github.com/relab/hotstuff/blob/crypto-refactor/.github/workflows/golangci-lint.yml#L17) with 1.18 would help.
